examples/merkle_proofs.md

```diff
@@ -0,0 +1,157 @@
+
+![The bank's data](./merkle_proof_1.jpg)
+
+### Producing Acccount Statements
+
+Let's produce an account statement for Eve.
+
+```ocaml
+let visit_tree tree =
+  (* [tree] is shallow. Let's only load the parts we are interested in *)
+  let+ (_ : int option) = Store.Tree.find tree [ "e"; "v"; "e" ] in
+  (Store.Tree.empty (), `Success)
+
+let proof, `Success = Lwt_main.run (
+  Store.Tree.produce_proof repo (`Node tree_key) visit_tree)
+
+let pp_merkle_proof = Irmin.Type.pp Store.Tree.Proof.tree_t
+```
+
+```ocaml
+# let () = Fmt.pr "proof: %a\n" pp_merkle_proof (Store.Tree.Proof.state proof);;
+proof: {"Node":[["b",{"Blinded_node":"584ba234368730ce6cea56d93433a7a777cad2a0"}],["e",{"Node":[["v",{"Node":[["e",{"Contents":[30,"normal"]}]]}]]}]]}
+```
+
+The proof of Eve's balance is a Merkle proof. It is a subset of the tree of all the accounts in the bank, it doesn't leaks Bob and Ben's balance, it only reveals the hash and the name of their subtree (i.e. `["b",{"Blinded_node":"584ba234368730ce6cea56d93433a7a777cad2a0"}]`).
+
+![The Merkle proof](./merkle_proof_2.jpg)
+
+The reason why this statement is temper proof is because it hashes the same way as the full account tree of the bank. The bank could make public the hash in a trusted place (e.g. on Twitter) and anyone could produce the hash of the statement and make sure that it is identical to the hash published by the bank.
+
+```ocaml
+let pp_hash = Irmin.Type.pp Store.hash_t
+```
+
+```ocaml
+# let () =
+    let bank_hash = Store.Tree.hash tree in
+    Fmt.pr " bank hash: %a\n" pp_hash bank_hash
+
+  let () =
+    let proof_hash =
+      match Store.Tree.(Proof.state proof |> hash_of_proof_state) with
+      | `Node h -> h
+      | `Contents _ -> assert false
+    in
+    Fmt.pr "proof hash: %a\n" pp_hash proof_hash
+  ;;
+ bank hash: 8e2d5291a09b98725e8525807d189eb7d9eb66d5
+proof hash: 8e2d5291a09b98725e8525807d189eb7d9eb66d5
+```
+
+### What `produce_proof` does
+
+Here is the signature of `produce_proof`.
+
+```ocaml
+# let produce_proof = Store.Tree.produce_proof;;
+val produce_proof :
+  Store.repo ->
+  Store.Tree.kinded_key ->
+  (Store.tree -> (Store.tree * 'a) Lwt.t) ->
+  (Store.Tree.Proof.tree Store.Tree.Proof.t * 'a) Lwt.t = <fun>
+```
+
+`produce_proof repo key_before f` is `(proof = { state; hash_before; hash_after }, f_res)`. `f` is invoked once per call to `produce_proof` and `f tree_before` is `(tree_after, f_res)`.
+
+`produce_proof` invokes `f`, passing it `tree_before` which is the tree at `key_before` in `repo`. In the preceding example, `tree_before` is the tree of all the bank accounts and `key_before` is the pointer in `repo` to the root of all the bank accounts.
+
+`f` is expected to perform operations on `tree_before`. Irmin will record all these operations and remember which paths were loaded and which paths weren't. All operations in `Store.Tree` work! All of this is thread safe because behind the scene, `tree_before` is decorated with caches.
```

Comment:
"All of this is thread safe because behind the scene, `tree_before` is decorated with caches."
Several questions about this claim:
- What is the meaning of "thread safe"?
- What is the caches content?
- How is the above claim sustained?